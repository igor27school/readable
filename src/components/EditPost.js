import React, { Component } from 'react'
import { connect } from 'react-redux'
import serializeForm from 'form-serialize'
import { Link } from 'react-router-dom'
import { fetchPostFromServer, editPost } from '../actions/ActionCreators'

class EditPost extends Component {
  componentDidMount() {
    const { post, postId, fetchPostFromServer } = this.props
    if (!post) {
      fetchPostFromServer(postId)
    }
  }
  handleSubmit = (e) => {
    e.preventDefault()
    const { post } = this.props
    const values = serializeForm(e.target, { hash: true })
    this.props.editPost(post.id, values)
    this.props.history.push(`/${post.category}/${post.id}`)
  }
  render() {
    const { post, postId } = this.props
    if (!post){
      return (
        <div>Post id is invalid or the post has been deleted: {postId}</div>
      )
    } else if (post.deleted) {
      return (
        <div>This post has been deleted: {postId}</div>
      )
    }
    return (
      <div>
        <Link to={`/${post.category}/${post.id}`}>Close</Link>
        <form onSubmit={this.handleSubmit}>
          <input type="text" name="title" defaultValue={post.title}/>
          <textarea name="body" defaultValue={post.body}></textarea>
          <button>Submit Edits</button>
        </form>
      </div>

    )
  }
}

function mapStateToProps ({ posts }, { match }) {
  return {
    post: posts.byId[match.params.post_id],
    postId: match.params.post_id,
  }
}

function mapDispatchToProps (dispatch) {
  return {
    fetchPostFromServer: (postId) => dispatch(fetchPostFromServer(postId)),
    editPost: (postId, values) => dispatch(editPost(postId, values)),
  }
}

export default connect(mapStateToProps, mapDispatchToProps)(EditPost)
