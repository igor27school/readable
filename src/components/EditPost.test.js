import React from 'react'
import ShallowRenderer from 'react-test-renderer/shallow'
import { EditPost } from './EditPost'

jest.mock('../utils/ServerAPI')

describe('EditPost tests', () => {
  const setup = () => {
    const props = {
      post: {
        id: 'testPostId',
        body: 'testing edit post',
        category: 'testCategory',
        deleted: false,
      },
      postId: 'testPostId',
      fetchPostFromServer: jest.fn(),
    }
    const renderer = new ShallowRenderer()
    return {
      props,
      renderer
    }
  }

  it('EditPost matches the snapshot', () => {
    const { renderer, props } = setup()
    renderer.render(<EditPost {...props}/>)
    const result = renderer.getRenderOutput()
    expect(result).toMatchSnapshot()
  })
})
