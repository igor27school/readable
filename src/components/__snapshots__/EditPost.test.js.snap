// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`EditPost tests EditPost matches the snapshot 1`] = `
<div>
  <Link
    replace={false}
    to="/testCategory/testPostId"
  >
    Close
  </Link>
  <form
    onSubmit={[Function]}
  >
    <input
      defaultValue={undefined}
      name="title"
      type="text"
    />
    <textarea
      defaultValue="testing edit post"
      name="body"
    />
    <button>
      Submit Edits
    </button>
  </form>
</div>
`;
